external/esp_idf_port/esp32/esp32_wifi_utils.c

```diff
@@ -0,0 +1,737 @@
+		g_cbk.sta_connected(WIFI_UTILS_SUCCESS, NULL);
+		ESP_LOGI(TAG, "[%s] SYSTEM_EVENT_STA_CONNECTED", __func__);
+		break;
+	}
+	case SYSTEM_EVENT_STA_DISCONNECTED: {
+		s_retry_num++;
+		int reason = event->event_info.disconnected.reason;
+		ESP_LOGI(TAG, "[%s] connect to the AP %s fail %d: %d; %d\n", __func__, event->event_info.disconnected.ssid, s_retry_num, event->event_info.disconnected.reason, isStaConnected);
+
+		if (s_retry_num > 0) {
+			ESP_LOGI(TAG, "g_cbk.sta_connected fail\n");
+			g_cbk.sta_connected(WIFI_UTILS_FAIL, NULL);
+			s_retry_num = 0;
+		} else {
+			ESP_LOGI("g_cbk.sta_disconnected\n");
+			g_cbk.sta_disconnected(NULL);
+		}
+		isStaConnected = 0;
+		break;
+	}
+	case SYSTEM_EVENT_AP_START: {
+		isStaConnected = 0;
+		ESP_LOGI(TAG, "ap start");
+		break;
+	}
+	case SYSTEM_EVENT_AP_STOP: {
+		ESP_LOGI(TAG, "ap stop");
+		break;
+	}
+	case SYSTEM_EVENT_AP_STACONNECTED: {
+		isAPJoined = 1;
+		ESP_LOGI(TAG, "station: join, AID=%d", MAC2STR(event->event_info.sta_connected.mac), event->event_info.sta_connected.aid);
+		break;
+	}
+	case SYSTEM_EVENT_AP_STADISCONNECTED: {
+		ESP_LOGI(TAG, "station: leave, AID=%d", MAC2STR(event->event_info.sta_disconnected.mac), event->event_info.sta_disconnected.aid);
+		g_cbk.softap_sta_left(NULL);
+		break;
+	}
+	case SYSTEM_EVENT_AP_STAIPASSIGNED: {
+		ESP_LOGI(TAG, "ap ip assigned");
+		g_cbk.softap_sta_joined(NULL);
+		break;
+	}
+	default: {
+		ESP_LOGI(TAG, "EVENT %d ", event->event_id);
+		break;
+	}
+	}
+	return ESP_OK;
+}
+
+//
+// Interface API
+//
+wifi_utils_result_e wifi_utils_init(void)
+{
+	esp_err_t ret;
+	tcpip_adapter_init();
+
+	s_retry_num = 0;
+	if (!event_init) {
+		ret = esp_event_loop_init(event_handler, NULL);
+		if (ret != ESP_OK) {
+			ESP_LOGI(TAG, "esp_event_loop_init failed, %d\n", ret);
+			return WIFI_UTILS_FAIL;
+		}
+		event_init = true;
+
+	}
+	esp_wifi_internal_set_log_level(WIFI_LOG_VERBOSE);
+	esp_wifi_internal_set_log_mod(WIFI_LOG_MODULE_ALL, WIFI_LOG_MODULE_ALL | WIFI_LOG_SUBMODULE_INIT | WIFI_LOG_SUBMODULE_IOCTL | WIFI_LOG_SUBMODULE_CONN | WIFI_LOG_SUBMODULE_SCAN, true);
+
+	if (!wifi_init) {
+		wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
+		ret = esp_wifi_init(&cfg);
+		if (ret != ESP_OK) {
+			ESP_LOGI(TAG, "esp_wifi_init failed, %d\n", ret);
+			return WIFI_UTILS_FAIL;
+		}
+		wifi_init = true;
+	}
+
+	ret = esp_wifi_set_storage(WIFI_STORAGE_RAM);
+	if (ret != ESP_OK) {
+		ESP_LOGI(TAG, "esp_wifi_set_storage failed, %d\n", ret);
+		return WIFI_UTILS_FAIL;
+	}
+
+	wifi_config_t wifi_config = {
+		.sta = {
+			.ssid = CONFIG_WIFI_SSID,
+			.password = CONFIG_WIFI_PASSWORD,
+			.scan_method = WIFI_FAST_SCAN,
+			.sort_method = WIFI_CONNECT_AP_BY_SIGNAL,
+			.threshold.rssi = -127,
+			.threshold.authmode = WIFI_AUTH_OPEN,
+		}
+		,
+	};
+
+	ret = esp_wifi_set_mode(WIFI_MODE_STA);
+	if (ret) {
+		ESP_LOGI(TAG, "esp_wifi_set_mode failed, %d\n", ret);
+		return;
+	}
+
+	ret = esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config);
```

Comment:
during wifi init,  wifi mode and config is set here.  any special reason here?
it seems that this shall be done in *_start_sta  or *_start_softap.

Reply:
yes, defaulty wifi is enter station mode in some utc and wm_test test.